lib/mongo/lib/dispatchers.js

```diff
@@ -1,82 +1,86 @@
-import compensateForLatency from './compensateForLatency';
-import { Events, RedisPipe } from '../../constants';
-import publish from './publish';
 import { Meteor } from 'meteor/meteor';
-import { Random } from 'meteor/random';
+import { DDPServer } from 'meteor/ddp-server';
+import { EJSON } from 'meteor/ejson';
+import { Events, RedisPipe } from '../../constants';
+import RedisSubscriptionManager from '../../redis/RedisSubscriptionManager';
+import { getRedisPusher } from "../../redis/getRedisClient";
 
-const dispatchUpdate = function(
-    optimistic,
-    collectionName,
-    channels,
-    docIds,
-    fields
-) {
-    if (optimistic) {
-        docIds.forEach(docId => {
-            compensateForLatency(
-                channels,
-                collectionName,
-                Events.UPDATE,
-                { _id: docId },
-                fields
-            );
-        });
+const getWriteFence = function (optimistic) {
+    if (optimistic && DDPServer._CurrentWriteFence) {
+        return DDPServer._CurrentWriteFence.get();
     }
-
-    Meteor.defer(() => {
-        docIds.forEach(docId => {
-            publish(
-                collectionName,
-                channels,
-                {
-                    [RedisPipe.EVENT]: Events.UPDATE,
-                    [RedisPipe.FIELDS]: fields,
-                    [RedisPipe.DOC]: { _id: docId }
-                },
-                docId
-            );
-        });
-    });
+    return null;
 };
 
-const dispatchRemove = function(optimistic, collectionName, channels, docIds) {
-    if (optimistic) {
-        docIds.forEach(docId => {
-            compensateForLatency(channels, collectionName, Events.REMOVE, {
-                _id: docId
+const dispatchEvents = function (fence, events, channels) {
+    if (fence) {
+        const write = fence.beginWrite();
+        RedisSubscriptionManager.queue.queueTask(() => {
+            events.forEach(({ event, dedicatedChannel }) => {
+                channels.forEach(channelName => {
+                    RedisSubscriptionManager.process(channelName, event);
+                });
+                RedisSubscriptionManager.process(dedicatedChannel, event);
             });
+            write.committed();
         });
     }
 
     Meteor.defer(() => {
```

Comment:
Another thing to take into account: https://github.com/cult-of-coders/redis-oplog/pull/257
Add that check as well for the external publisher.

We will soon open-source a go lang-written tool that would tail mongodb oplog and push to redis-oplog. I know that it sounds a bit strange, since we wanted to avoid mongo oplog all together, but it's actually very good when you have writes to db written from many languages and services, + you don't have to deal with race-conditions since mongo oplog operations come in order, meaning that we can say bye-bye to some of the db requests redis-oplog does to avoid this, and actually increasing performance even further.

Reply:
Yep, I'll rebase atop upstream master (& add re-implement mods)